### PR TITLE
opt:opt ltor masks

### PR DESCRIPTION
**Pull request by Baibaifan**
**Problem:**

In Megatron-LM, there is a memory bottleneck when using the reset attention mask to construct long sequences. The following code: (_get_ltor_masks_and_position_ids)
image

When a seq_len consists of multiple short documents, there will be multiple values ​​in eod_index. Each value means that attention_mask needs to be accessed and loaded once, and the corresponding position is assigned 0. For example, in the 32k scenario, there are multiple assignments in extreme scenarios, which makes the data loading time very slow. As the sequence length increases, the number of positions that need to be assigned increases, and the time consumption will be longer. As shown in the figure below.
image

**Solution**
Perform a tensor access using the  `block_diag` value.
image

### CODE REVIEW

**Comment by sbhavani** [comment]
@Baibaifan Thanks for your contribution on optimizing ltor masks for long sequences! As part of our new year cleanup, we're closing PRs opened before 2025.

This performance optimization for reset_attention_mask may still be relevant @NVIDIA/megatron-oncall

If you'd like to pursue this, please feel free to submit a fresh PR. Thanks for your understanding!
